pkg/rbac/common.go

```diff
@@ -0,0 +1,64 @@
+package rbac
+
+import (
+	"github.com/pkg/errors"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+/*BuildSubjectFromRTB This function will generate
+PRTB and CRTB to the subject with user or group
+and generate APP to the subject with ServiceAccount
+*/
+func BuildSubjectFromRTB(object interface{}) (rbacv1.Subject, error) {
+	var userName, groupPrincipalName, groupName, name, kind, saName, namespace string
+	if rtb, ok := object.(*v3.ProjectRoleTemplateBinding); ok {
+		userName = rtb.UserName
+		groupPrincipalName = rtb.GroupPrincipalName
+		groupName = rtb.GroupName
+		namespace = rtb.ServiceAccountNamespace
+		saName = rtb.ServiceAccountName
+	} else if rtb, ok := object.(*v3.ClusterRoleTemplateBinding); ok {
+		userName = rtb.UserName
+		groupPrincipalName = rtb.GroupPrincipalName
+		groupName = rtb.GroupName
+	} else {
+		return rbacv1.Subject{}, errors.Errorf("unrecognized roleTemplateBinding type: %v", object)
+	}
+
+	if userName != "" {
+		name = userName
+		kind = "User"
+	}
+
+	if groupPrincipalName != "" {
+		if name != "" {
+			return rbacv1.Subject{}, errors.Errorf("roletemplatebinding has more than one subject fields set: %v", object)
+		}
+		name = groupPrincipalName
+		kind = "Group"
+	}
+
+	if groupName != "" {
+		if name != "" {
+			return rbacv1.Subject{}, errors.Errorf("roletemplatebinding has more than one subject fields set: %v", object)
+		}
+		name = groupName
+		kind = "Group"
+	}
+
+	if namespace != "" {
```

Comment:
Would be better to check `saName` instead of `namespace` here. It is similar but more readable.